### PR TITLE
kernel/hil/time: allow time to be a dyn trait object

### PR DESCRIPTION
**Pull request by jettr**
### Pull Request Overview

Move the tick conversion functions to a separate auto-implemented
trait. This allows `dyn` on `Time`, `Alarm`, and `Timer` which is helpful since
these traits are normally abstracted with a VirtualMux.

Add the inverse conversion of tick to time functions as well.

### Testing Strategy

Added unit tests

### Documentation Updated

- Updated time TRD in PR #2688 

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

**Review by brghena — COMMENTED** [review]
Someone else will have to speak to whether there are any runtime impacts here, but in my opinion this makes the code a lot cleaner.

File: capsules/src/test/random_alarm.rs
Comment (jettr) [reply capsules/src/test/random_alarm.rs:22]: We could change it to `alarm.ticks_from_seconds(0)`, however I do think that the assumption that 0 is the same will always hold true since tick is just another scaled measure of time, and 0 * any_scale => 0

File: capsules/src/virtual_alarm.rs
Comment (jettr) [reply capsules/src/virtual_alarm.rs:37]: I will update all the references together depending on what we decide. There are a few :)

File: kernel/src/hil/time.rs
Comment (jettr) [reply kernel/src/hil/time.rs:629]: We can only return a u32, so we can't use u64:MAX.

The tick conversion methods before assumed we didn't want to deal with more than u32 for time (which I think is a fine assumption), so I held that assumption in the new conversion methods. We limit all conversions to saturated u32.

File: kernel/src/hil/time.rs
Comment (jettr) [inline kernel/src/hil/time.rs:698]: It is also worth nothing that this test would have failed before this commit, the result would have been `u64::MAX` instead of `1_000_000u64 << 31`

File: kernel/src/hil/time.rs
Comment (brghena) [reply kernel/src/hil/time.rs:629]: I understand, the `u32` returned here is only ever in units of time, rather than units of ticks. I think this does make sense then. (Whether time should be specifiable in u64s is a possible discussion, but outside of the scope of this PR.)

File: capsules/src/test/random_alarm.rs
Comment (brghena) [reply capsules/src/test/random_alarm.rs:22]: I don't feel so strongly about this. Personally, I'd probably write `self.alarm.ticks_from_seconds(0)` to reinforce what the number signifies, but your logic makes sense too.

Unless anyone else does feel strongly and wants to follow up, this can probably stay as-is.

**Review by hudson-ayers — APPROVED** [review]
I like the method syntax much more than the associated function syntax for calling these helper functions, this is great.

I verified it has no impact on code size. While this technically adds an argument to each of these functions (a reference to self), I suspect the additional argument is optimized out as a result on being unused in most cases.

While there is of course overhead to using Time as a dyn trait object instead of a generic parameter, I still think we should make it possible it to use it as such if someone has the need.

**Review by brghena — APPROVED** [review]

File: kernel/src/hil/time.rs
Comment (lschuermann) [inline kernel/src/hil/time.rs:55]: I don't understand this part. This function is supposed to, for time `T: Ticks`, given a `u64`, construct a `T` with either the passed number or `<T as Ticks>::max_value()`, right?

Won't comparing with `Self::max_value().into_u32()` set the upper bound to a `u32` then?

File: capsules/src/test/random_alarm.rs
Comment (lschuermann) [reply capsules/src/test/random_alarm.rs:22]: I agree with @brghena. I agree that @jettr's logic is correct, however `self.alarm.ticks_from_seconds(0)` makes it much more obvious that we're talking about some `Ticks` here. The `Into` trait is good for when it's obvious what type the conversion will result in, but arguable here it's not the case. I'd also be okay with `Ticks::from(0u32)`.

**Review by lschuermann — COMMENTED** [review]
These changes look good to me in general.

I think we're going to want to be careful as to where we are using trait objects, as this might come with drastic increases in code size and significantly reduces optimization potential. As @hudson-ayers evaluated, it seems fine for the cases here.

**Comment by phil-levis** [comment]
Because time is specified in a finalized TRD, these changes would require a new TRD. The intention in the time system design had been to avoid the use of dyn. 

Phil [sent from a phone]

**Comment by lschuermann** [comment]
For reference, this is the document Phil is talking about:
doc/reference/trd101-time.md

Philip Levis ***@***.***> writes:
> Because time is specified in a finalized TRD, these changes would
> require a new TRD.

Correct me if I'm wrong: if these changes were entirely compatible with
the current TRD, they wouldn't require any changes or a new TRD, right?

Having read over the entire document, I don't think these changes affect
the overall structure and intention of what's specified. However, they
would cause a mismatch between the code snippets in the TRD and the
proper implementation.

I expect to run into these issues more than once: what's the policy
here? Is even a slight mismatch between the code snippets and actual
implementation enough grounds to require a new TRD. Further, do we have
a deperaction process for TRDs if one were inclined to change it as
proposed here?

> The intention in the time system design had been to avoid the use of
> dyn.

Agreed. I do think that these changes as proposed here seem fine, given
that they appear to be optimized out entirely, as @hudson-ayers
stated. However, we'd have to be very careful as treating Time as a
trait object might have an impact if we use it differently, or add more
complex methods in the future which make use of the &self parameter.

**Comment by phil-levis** [comment]
> I expect to run into these issues more than once: what's the policy here? Is even a slight mismatch between the code snippets and actual implementation enough grounds to require a new TRD. Further, do we have a deprecation process for TRDs if one were inclined to change it as proposed here?

It requires a new TRD. I think it would require more than just changing the code snippets. It would require adding the `ConvertTicks` trait (since it's part of the API).

TRD1 describes the process: we would amend the TRD, create a new TRD with a new number, and add an "Obsoleted By" field to TRD101:

```The "Obsoletes" field is a backward pointer to an earlier TRD which
the current TRD renders obsolete. An Obsoletes field MAY have multiple
TRDs listed.  For example, if TRD 121 were to replace TRDs 111 and
116, it would have the field "Obsoletes: 111, 116".

The "Obsoleted By" field is added to a Final TRD when another TRD has
rendered it obsolete. The field contains the number of the obsoleting
TRD. For example, if TRD 111 were obsoleted by TRD 121, it would have
the field "Obsoleted By: 121".

"Obsoletes" and "Obsoleted By" fields MUST agree. For a TRD to list another
TRD in its Obsoletes field, then that TRD MUST list it in the Obsoleted By
field.

The obsoletion fields are used to keep track of evolutions and modifications
of a single abstraction. They are not intended to force a single approach or
mechanism over alternative possibilities.
```

This process is intended to be cumbersome on purpose: once something is finalized, we should avoid changing it as much as we can. The core abstractions of an operating system should not have churn. That being said, there are times when changes are needed, which is why this mechanism is in place. 

I'm personally against this change unless there's a compelling specific use case. I'm conservative about these things. E.g., if there's a situation where we really *have* to use `dyn`, or there are significant disadvantages to not using `dyn`, then I'm all for it. But lots of little improvements spread over the kernel. are a kind of Brownian motion that makes it very difficult to keep branches synced to master.

File: capsules/src/virtual_timer.rs
Comment (jettr) [reply capsules/src/virtual_timer.rs]: Done.

File: kernel/src/hil/time.rs
Comment (jettr) [reply kernel/src/hil/time.rs:55]: This provided implementation works for all `Ticks` expect `Ticks64`, which provides its own implementation. I could instead copy and paste this same code as the implementation for `Ticks16`, `Ticks24`, and `Ticks32` and not have the default implementation.

Even though I lean toward keeping this default implementation for everything but `Ticks64`, I can remove it in favor of every `Ticks` implementing explicitly. Just let me know.

**Comment by jettr** [comment]
I am sure this isn't the correct procedure, but I updated `trd101-time.md` to reflect the changes that this commit would bring. If the text changes seem okay, then I can make a new trd file with all of the content from the existing trd101-time along with these changes and obsolete the current one (if that is the correct way to do it)

This PR is still compatible with all of the tenants and fundamentals that the trd101-time.md has already established. This PR just adds more functionality than was previously available for down stream users of this code. In the trd edits, I still call out that using generic arguments for `Time` and sub traits is still the prefer methodology (for good reason). It seems that the prevention of trait objects on `Time` was merely accidental instead of intentional. I think we should leave it up to the down stream consumer to make the trait object versus generic parameter trade-off.

Another factor that I really like about this change comes into play if there are two different alarms present in the same driver. It is much more clear you are performing the correct tick conversion if you are using the same object that you set the alarm with in the next line. You don't have to trace the generic parameter back to the correct alarm field to know that it is correct.

**Comment by phil-levis** [comment]
It sounds like will hopefully discuss this on Friday's call.

**Comment by phil-levis** [comment]
Looks like this needs to sync with master before I review?

**Comment by jettr** [comment]
> Looks like this needs to sync with master before I review?

Done. It is up to date now.

File: kernel/src/hil/time.rs
Comment (lschuermann) [reply kernel/src/hil/time.rs:55]: Can we please either add a warning comment that this auto-implemented function will not work for any Ticks > 32-bit, or just not provide an auto-implementation?

I have a bad feeling about this. An auto-implementation should always work if it's there. Not implementing this on a custom `Ticks` type is going to cause a ton of headaches. Also, people may easily miss the Ticks64 implementation of this method and assume wrong behavior, given that other Ticks don't have their own implementation of this.

A clearly visible comment might be fine, if an explicit implementation per `impl Ticks` feels to verbose for you.

Comment (jettr) [reply kernel/src/hil/time.rs]: Done. I just implemented in every type. This way there are not future landmine (even though having another ticks greater than 32 is very unlikely).

**Review by alistair23 — APPROVED** [review]

**Review by hudson-ayers — DISMISSED** [review]

**Review by phil-levis — DISMISSED** [review]
This needs to remove the TRD101 changes before it can be merged.

**Comment by jettr** [comment]
> This needs to remove the TRD101 changes before it can be merged.

Good call. Done and rebase to the latest commit.

**Review by hudson-ayers — APPROVED** [review]

**Pull request merged**
